documentation/general/dotnet-run-file.md

```diff
@@ -0,0 +1,266 @@
+>
+> If we want to also support [multi-entry-point scenarios](#multiple-entry-points),
+> we might need an option like `dotnet run --entry ./dir/name`
+> which would work for both `./dir/name.cs` and `./dir/name/name.csproj`.
+
+File-based programs are processed by `dotnet run` equivalently to project-based programs unless specified otherwise in this document.
+For example, the remaining command-line arguments after the first argument (the target path) are passed through to the target app
+(except for the arguments recognized by `dotnet run` unless they are after the `--` separator).
+
+## Entry points
+
+If a file is given to `dotnet run`, it has to be an *entry-point file*, otherwise an error is reported.
+We want to report an error for non-entry-point files to avoid the confusion of being able to `dotnet run util.cs`.
+
+Currently, entry-point files must contain top-level statements,
+but other entry-point forms like classic `Main` method can be recognized in the future.
+We could modify Roslyn to accept the entry-point path and then it would be the compiler's responsibility
+to check whether the file contains an entry point (of any kind) and report an error otherwise.
+
+Because of the [implicit project file](#implicit-project-file),
+other files in the target directory or its subdirectories are included in the compilation.
+For example, other `.cs` files but also `.resx` (embedded resources).
+Similarly, implicit build files like `Directory.Build.props` are used during the build.
+
+> [!NOTE]
+> Performance issues might arise if there are many [nested files](#nested-files) (possibly unintentionally),
+> and also it might not be clear to users that `dotnet run file.cs` will include other `.cs` files in the compilation.
+> Therefore we could consider some switch (a command-line option and/or a `#` language directive) to enable/disable this behavior.
+> When disabled, [grow up](#grow-up) would generate projects in subdirectories similarly to [multi-entry-point scenarios](#multiple-entry-points)
+> to preserve the behavior.
+
+### Nested files
+
+If there are nested project files like
+```
+App/File.cs
+App/Nested/Nested.csproj
+App/Nested/File.cs
+```
+executing `dotnet run app/file.cs` includes the nested `.cs` file in the compilation.
+That might be unexpected, hence we could consider reporting an error in such situation.
+However, the same problem exists for normal builds with explicit project files
+and usually the build fails because there are multiple entry points or other clashes.
+
+Similarly, we could report an error if there are many nested directories and files,
+so for example if someone puts a C# file into `C:/`
+and executes `dotnet run C:/file.cs` or opens that in the IDE, we do not walk the whole drive.
+Again, this problem exists with project-based programs as well, albeit it's likely uncommon there
+since creating projects with `dotnet new` or an IDE creates them in a directory.
```

Comment:
This isn't actually true by default for `dotnet new`, it puts the output in the current directory unless `-n` or `-o` are specified.